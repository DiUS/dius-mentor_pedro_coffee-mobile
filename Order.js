'use strict';

import React, { Component } from 'react';
import { View, ListView, Button, Text, TextInput, StyleSheet, ActivityIndicator } from 'react-native';
import OptionsListView from './OptionsListView';
import ProceedButton from './ProceedButton';
import ApiUtils from './ApiUtils';

class Order extends Component{
  constructor(props){
    super(props);
    this._initialName=this.props.order.name;
    this._initialDrinks=this.props.order.coffees;
    this._drinks = null;
    this.state = {
      name : this.props.order.name,
      order : this.props.order,
      drinks : null
    }
  }

  onCancel(){
    return fetch(ApiUtils.ENDPOINT+this.state.order.id,{
      method: 'DELETE',
      headers: {'Content-Type': 'application/json'}
    })
    .then(ApiUtils.checkStatus)
    .then((response) => response.json())
    .then((responseJson) => {
      this.props.onComplete();
    })
    .catch((error) => {
      console.error(error);
    })
  }

  onBack(){
    this.props.onComplete();
  }

  onSave(){
    return fetch(ApiUtils.ENDPOINT+this.state.order.id,{
      method: 'PATCH',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({name : this.state.name})
    })
    .then(ApiUtils.checkStatus)
    .then((response) => response.json())
    .then((responseJson) => {
      this.props.onComplete();
    })
    .catch((error) => {
      console.error(error);
    })
  }

  onSelectDrink(drink){
    this.props.onSelect(drink);
  }

  onDeleteDrink(drink){
    return fetch(ApiUtils.ENDPOINT+this.state.order.id+'/'+ApiUtils.COFFEE_PATH+drink.id,{
      method: 'DELETE',
      headers: {'Content-Type': 'application/json'}
    })
    .then(ApiUtils.checkStatus)
    .then((response) => response.json())
    .then((responseJson) => {
      var newDrinks = this._drinks.filter((item) =>{
        if(item.id !== drink.id){
          return item;
        }
      })
      this.updateState(newDrinks);
    })
    .catch((error) => {
      console.error(error);
    })
  }

  updateState(drinkList){
    var ds = new ListView.DataSource({rowHasChanged : (r1, r2) => r1 !== r2});
    this._drinks=drinkList;
    this.setState({
      drinks : ds.cloneWithRows(drinkList)
    });
  }

  componentDidMount(){
    var drinkList = [];
    var order = this.state.order;
    if(order.coffees?order.coffees.length===0:true){
      this.updateState(drinkList);
    }
    else{
      for(var i=0;i<order.coffees.length;i++){
        fetch(ApiUtils.ENDPOINT+order.id+'/'+ApiUtils.COFFEE_PATH+order.coffees[i].id)
        .then(ApiUtils.checkStatus)
        .then((response) => response.json())
        .then((responseJson) => {
          drinkList = [...drinkList,responseJson];
          if(drinkList.length===order.coffees.length){
            //TODO FIX THIS TO DO IT IN A CORRECT WAY... ASYNC? PROMISE?
              this.updateState(drinkList);
          }
        })
        .catch((error) => {
          console.error(error);
        })
      }
    }
  }

  _renderRowSelect(rowData){
    return (
      <View style={{flexDirection: 'row', justifyContent:'space-between'}}>
        <View style={{flex : 5}}>
          <Button title={rowData.size+' '+rowData.style} onPress={()=>this.onSelectDrink(rowData)}/>
        </View>
        <View style={{marginLeft: 4, flex : 1}}>
          <Button title={'X'} onPress={()=>this.onDeleteDrink(rowData)} color='#ff6666'/>
        </View>
      </View>
    );
  }

  render(){
    if(this.state.drinks){
      return(
        <View style={{flex: 1, padding: 10, justifyContent:'space-between'}}>

          <TextInput
            style={{height: 60, fontSize: 24, fontWeight:'bold'}}
            onChangeText={(name) => this.setState({name:name})}
            value={this.state.name}
          />

          <View style={{flexDirection:'row', justifyContent:'space-between', marginBottom:8}}>
              <View style={{flex : 5}}>
                <Text style={{fontSize: 24, fontWeight:'bold'}}>Drinks:</Text>
              </View>
              <View style={{marginLeft: 4, flex : 1}}>
                <Button title={'+'} onPress={()=>this.onSelectDrink(null)}/>
              </View>
          </View>

          <OptionsListView
            dataSource={this.state.drinks}
            renderRow={this._renderRowSelect.bind(this)}/>

          <View style={{flexDirection:'row', justifyContent:'space-between'}}>
            <View style={{flex : 1, marginRight:4}}>
              <Button title='<' onPress={()=>this.onBack()} disabled={!this.state.name || !this._initialName}/>
            </View>
            <View style={{flex : 4, marginLeft:4, marginRight:4}}>
              <Button title='Cancel Order' onPress={()=>this.onCancel()} color='#ff6666'/>
            </View>
            <View style={{flex : 4, marginLeft:4}}>
              <ProceedButton title='Save Order' onPress={()=>this.onSave()}
                disabled={(this.state.name===this._initialName) || (this.state.name==='')}/>
            </View>
          </View>

        </View>
      );
    }
    else {
      return(
        <View style={{flex: 1, padding: 10}}>
          <ActivityIndicator
            style={[styles.center, {height: 80}]}
            size="large"/>
        </View>
      );
    }
  }
}

const styles = StyleSheet.create({
  center: {
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
    padding: 8,
  },
});

export default Order;
