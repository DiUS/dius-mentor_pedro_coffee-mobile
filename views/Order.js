'use strict';

import React, { Component } from 'react';
import { View, ListView, Button, Text, TextInput, StyleSheet, ActivityIndicator } from 'react-native';
import OptionsListView from '../components/OptionsListView';
import Options from '../components/Options';

class Order extends Component{
  constructor(props){
    super(props);
    var ds = new ListView.DataSource({rowHasChanged : (r1, r2) => r1 !== r2});
    //Initial values needed to toggle action buttons
    this._initialName=this.props.order.name;
    this._drinks=this.props.order.coffees?this.props.order.coffees:[]
    this.state = {
      name : this.props.order.name,
      order : this.props.order,
      drinks : ds.cloneWithRows(this._drinks)
    }
  }

  onBack(){
    this.props.onUpdateOrder();
  }

  renderRowSelect(rowData){
    return (
      <View style={{flexDirection: 'row', justifyContent:'space-between'}}>
        <View style={{flex : 5}}>
          <Button title={rowData.summary} onPress={()=>this.props.onSelectDrink(rowData)}/>
        </View>
        <View style={{marginLeft: 4, flex : 1}}>
          <Button title={'X'} onPress={()=>this.props.onDeleteDrink(rowData)} color='#ff6666'/>
        </View>
      </View>
    );
  }

  renderLoading(){
    return(
      <View style={{flex:1}}>
        <ActivityIndicator
          style={styles.spinner}
          size="large"/>
      </View>
    );
  }

  renderOrder(){
    return(
      <View style={{flex:1}}>
        <TextInput
          style={[styles.text,{height: 60}]}
          onChangeText={(name) => this.setState({name:name})}
          value={this.state.name}
        />

        <View style={styles.containerRow}>
            <View style={{flex : 5}}>
              <Text style={styles.text}>Drinks:</Text>
            </View>
            <View style={{marginLeft: 4, flex : 1}}>
              <Button title={'+'} onPress={()=>this.props.onSelectDrink()}/>
            </View>
        </View>

        <OptionsListView
          dataSource={this.state.drinks}
          renderRow={this.renderRowSelect.bind(this)}/>

        <Options
          onBack={this.onBack.bind(this)}
          onBackDisabled={!this.state.name || !this._initialName}
          onCancel={()=>this.props.onDeleteOrder(this.state.order.id)}
          onSave={()=>this.props.onUpdateOrder(this.state.order.id,this.state.name)}
          onSaveDisabled={(this.state.name===this._initialName) || (this.state.name==='')}/>
      </View>
    );
  }

  render(){
    return (
      <View style={styles.container}>
        {this.state.drinks && this.renderOrder()}
        {!this.state.drinks && this.renderLoading()}
      </View>
    );
  }

}

const styles = StyleSheet.create({
  spinner: {
    alignItems: 'center',
    justifyContent: 'center',
    height: 80,
    flex: 1,
    padding: 8,
  },
  text:{
    fontSize: 24,
    fontWeight:'bold'
  },
  containerRow:{
    flexDirection:'row',
    justifyContent:'space-between',
    marginBottom:8
  },
  container:{
    flex: 1,
    padding: 10,
    justifyContent:'space-between'}
});

export default Order;
