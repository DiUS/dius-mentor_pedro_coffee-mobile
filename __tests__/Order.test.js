import 'react-native';
import React from 'react';
import Order from '../views/Order.js';
import renderer from 'react-test-renderer';
import 'isomorphic-fetch';

var order = {
  id:1,
  coffees:null,
  name:"test_order",
  path:"order/1"
}
it('renders correctly', () => {
  const instance = renderer.create(
    <Order order={order}/>
  );
  expect(instance.toJSON()).toMatchSnapshot();
});
