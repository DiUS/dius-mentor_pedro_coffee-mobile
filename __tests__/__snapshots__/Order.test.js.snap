// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      style={
        Array [
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
          },
          Object {
            "height": 60,
          },
        ]
      }
      value="test_order"
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 8,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 5,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          Drinks:
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "marginLeft": 4,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#0C42FD",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              +
            </Text>
          </View>
        </View>
      </View>
    </View>
    <RCTScrollView
      dataSource={
        ListViewDataSource {
          "items": 0,
        }
      }
      enableEmptySections={true}
      renderRow={[Function]}
      renderScrollComponent={[Function]}
      renderSeparator={[Function]}
    >
      <View />
    </RCTScrollView>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 4,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#0C42FD",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              &lt;
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 4,
            "marginLeft": 4,
            "marginRight": 4,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#0C42FD",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#ff6666",
                  },
                ]
              }
            >
              Cancel
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 4,
            "marginLeft": 4,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
              "disabled",
            ]
          }
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {},
                Object {},
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#0C42FD",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                  Object {
                    "color": "#79d279",
                  },
                  Object {
                    "color": "#cdcdcd",
                  },
                ]
              }
            >
              Save
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
